### ISSUE TITLE
insert row does not use default when function

### ISSUE THREAD

**Issue by mtissington** [sqlitebrowser/sqlitebrowser#954]
CREATE TABLE `test2` (
	`id`	TEXT NOT NULL DEFAULT (randomblob(6)),
	`name`	INTEGER,
	PRIMARY KEY(`id`)
);

Browse data and the click new record, default has wrong value.

**Comment by justinclift** [comment]
Interesting.  Which version is this happening with, and on which OS? :smile:

If you're on Windows or OSX, would you be ok to try the latest nightly build? (download locations are listed on our front page)

Asking because we've improving our constraint and foreign key handling code recently, so there's a decent chance this is already fixed. 😁

**Comment by mtissington** [comment]
Its on Windows with the latest nightly build .99

**Comment by justinclift** [comment]
@MKleusberg This issue sounds like it'll be of interest to you.

@mtissington Thanks. :smile: The .99 bit is kind of placeholder text to signal "this is a development build".  We really need to get that updated at some point so development builds shows the git commit hash they were built from instead (much more useful). :wink:

**Comment by chrisjlocke** [comment]
> We really need to get that updated at some point so development builds shows the git commit hash they were built from instead

Or the date of the build.... that must be really simple to do?

**Comment by MKleusberg** [comment]
This is actually easy to fix but that fix comes with a caveat. Suppose having this table layout:
```sql
CREATE TABLE test(id integer NOT NULL DEFAULT 1, value text, PRIMARY KEY(id));
```
In this case I'm not so sure we want to use the default value because then the 'Add record' button stops working after the first insert because you can't use the value 1 multiple times. I don't think that's what users expect for table schemas like that either. So we're in a dilemma here.

I think fetching any duplicate errors from SQLite and then falling back to the old method isn't an option because it's really insecure and ugly programming-wise. And @innermous would hate us for doing that (see issue #628) :wink: So we'd have to parse the default value and see if it's a ```random*()``` function call or something like that. But that's not so easy either.

So I'm a bit in a loss here :frowning: Any other ideas?

**Comment by MKleusberg** [comment]
@chrisjlocke Build date sounds good. We used to have the git commit hash in there but it never really worked that well. I don't remember the reason why we deleted it but I do remember how I happy I was to finally get rid of all the build script magic that was needed :wink:

**Comment by justinclift** [comment]
@MKleusberg You sure we ever got it working (the git hash thing)?  /me really doesn't remember it :wink:

**Comment by justinclift** [comment]
> Or the date of the build.... that must be really simple to do?

Well, I guess it depends on what it would take to get a useful string (date, git commit hash, etc?) turned into a string that can be compiled in at build time.

If we need to use OS specific stuff (eg sh on *nix, batch file stuff on Win) it could get tricky.

**Comment by MKleusberg** [comment]
@justinclift See commit 5c1347bb927e72227af1e744f758f5dd07adf75a, 441d4e79b7d4f8cac4256b6f9369045ecabe3338 and 69a6685eaae0bbf33e76338c309f1e34170b08c9 for example. We definitely tried to get it working :smiley: Seems like it only ever worked using the cmake build though.

**Comment by chrisjlocke** [comment]
> Well, I guess it depends on what it would take to get a useful string [...] turned into a string that can be compiled in at build time
Does it have to though?  In VB, 'it' knows details about itself (version, current directory, etc) and as such its full path.  This is passed to the IO to determine the last modification date of said file...
Dont' know if any of that can be cross-platform.....  *mind boggles, eyes glaze*

**Comment by justinclift** [comment]
@MKleusberg Thanks. :smile:

**Comment by justinclift** [comment]
@chrisjlocke Yep, there might be compiler or build system variables we could use.  That would be good. :smile:

**Comment by mtissington** [comment]
If the user (me) is intelligent enough to create a default value that is random (and possibly a constraint) then do you need to check for it causing duplicates?

If you do, instead of trying to parse it and check if it is random can you do an if exists first?

**Comment by MKleusberg** [comment]
@mtissington The random value is fine. Especially in your case collisions aren't very likely either. But what I'm worried about is that the same code (on our side) also had to handle non-random default values that are used in primary key or unique columns.

So I'm wondering what sort of logic I could implement to support your case (maybe even ignoring collisions) but also to not screw it up for non-random default values :smiley:
